### PR TITLE
Update skeleton

### PR DESCRIPTION
**Pull request by carector**
Update implementation for StringManipulation and StringManipulationTest

### CODE REVIEW

**Comment by bpsix** [comment]
Overall looks good. You use the split() method throughout, checking for newlines and space characters. I also did this, but apparently \s covers tabs, newlines, carriage returns and a few more characters. So, like my code, the regex can be replaced with just "\s".

In your restoreString() method, I like how you used 1 loop to simultaneously check for invalid indices and, if they are valid, just update the to-be returned string. I opted for two loops, checking for valid indices first, but your solution is more efficient.

One thing that may help in the future is adding more comments so your thought process can be followed with ease.

**Review by bpsix — APPROVED** [review]

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,33 +1,89 @@
...
+        s = string;
     }
 
     @Override
     public int count() {
-        return 0;
+        if(s.length() == 0)
+            return 0;
+
+        String[] arr = s.split("[\\s\\n]");
```
Comment (bpsix) [inline src/main/java/StringManipulation.java:30]: Test

Comment (bpsix) [inline src/main/java/StringManipulation.java:30]: This can be simplified as: String[] arr = s.split("\\\s+")

There should be a '+' in case there are multiple spaces in between words

File: src/test/java/StringManipulationTest.java [hunk 2]
```diff
@@ -26,19 +36,28 @@ public void testCount1() {
...
     }
 
+    // Tests number of words in string with newline characters
     @Test
     public void testCount4() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("Including\nsome\nnewline\ncharacters");
+        int length = manipulatedstring.count();
+        assertEquals(4, length);
     }
```
Comment (bpsix) [inline src/test/java/StringManipulationTest.java:61]: Have you considered adding a test case that checks the null string case?

File: src/test/java/StringManipulationTest.java [hunk 3]
```diff
@@ -50,100 +69,194 @@ public void testRemoveNthCharacter1() {
...
+        manipulatedstring
+                .setString("Special characters! Punctuation, symbols !@#$ and numbers 91283... Will it matter?");
+        String[] substrings = manipulatedstring.getSubStrings(2, 10);
+        assertEquals(substrings[0], "characters!");
+        assertEquals(substrings[3], "!@#$");
+        assertEquals(substrings[6], "91283...");
     }
+
+    // Test whether exceptions are thrown when startWord > endWord or when one of the variables is < 0
     @Test
```
Comment (bpsix) [inline src/test/java/StringManipulationTest.java:159]: In Java 8+, you can use Lamba expressions with JUnit to handle exceptions. Here's a resource that shows how this is done: https://blog.codeleak.pl/2014/07/junit-testing-exception-with-java-8-and-lambda-expressions.html

File: src/test/java/StringManipulationTest.java [hunk 4]
```diff
@@ -50,100 +69,194 @@ public void testRemoveNthCharacter1() {
...
+        try {
+            manipulatedstring.getSubStrings(0, -3);
+            fail("No exception thrown");
+        } catch (IllegalArgumentException e) {
+        }
     }
+
+    // Tests whether an exception is thrown when endWord is out of bounds
     @Test
     public void testGeSubStrings6() {
```
Comment (bpsix) [inline src/test/java/StringManipulationTest.java:184]: See line 160 comment

File: src/test/java/StringManipulationTest.java [hunk 5]
```diff
@@ -50,100 +69,194 @@ public void testRemoveNthCharacter1() {
...
+        try {
+            manipulatedstring.removeNthCharacter(0, false);
+            fail("No exception thrown for n=0");
+        } catch (IllegalArgumentException e) {
+            assert true;
+        }
     }
 
+    // Tests whether method throws error when n > length
     @Test
```
Comment (bpsix) [inline src/test/java/StringManipulationTest.java:109]: I shared a link on line 160 that talks about using Lamba expressions when dealing with exceptions in JUnit.

File: src/test/java/StringManipulationTest.java [hunk 6]
```diff
@@ -50,100 +69,194 @@ public void testRemoveNthCharacter1() {
...
-    public void testRestoreString2()
-    {
-        fail("Not yet implemented");
-
+    public void testRestoreString2() {
+        manipulatedstring.setString("calebrector");
+        int[] array = new int[] {0, 5, 1, 4, 3, 2, 6, 7, 8, 9, 10};
+        String restoreString = manipulatedstring.restoreString(array);
+        assertEquals(restoreString, "crabelector");
```
Comment (bpsix) [inline src/test/java/StringManipulationTest.java:209]: Lol, crabelector

File: src/test/java/StringManipulationTest.java [hunk 7]
```diff
@@ -50,100 +69,194 @@ public void testRemoveNthCharacter1() {
...
-
+    public void testRestoreString4() {
+        manipulatedstring.setString("CAPStest");
+        int[] array = new int[] {4, 5, 6, 7, 0, 1, 2, 3};
+        String restoreString = manipulatedstring.restoreString(array);
+        assertEquals(restoreString, "testCAPS");
     }
 
+    // Tests whether exceptions are thrown when an index is negative or out of bounds
     @Test
```
Comment (bpsix) [inline src/test/java/StringManipulationTest.java:231]: See line 160 comment. It would help to split this test case into several smaller test cases.

File: src/test/java/StringManipulationTest.java (hunk 3 above)
Comment (carector) [reply src/test/java/StringManipulationTest.java:159]: Thanks! I've been meaning to work more with lambda expressions. I'll be sure to check out your implementation as well.
